util/stmtsummary/statement_summary.go

```diff
@@ -383,25 +405,68 @@ func (ssMap *stmtSummaryByDigestMap) SetRefreshInterval(value string, inSession
 			interval = 0
 		}
 	}
-	if interval > 0 {
-		atomic.StoreInt64(&ssMap.sysVars.refreshInterval, int64(interval))
+	// Users may set invalid values in configuration file.
+	if interval <= 0 {
+		interval = 1
 	}
+	atomic.StoreInt64(&ssMap.sysVars.refreshInterval, int64(interval))
 }
 
-func (ssMap *stmtSummaryByDigestMap) RefreshInterval() int64 {
+// refreshInterval gets the refresh interval for summaries.
+func (ssMap *stmtSummaryByDigestMap) refreshInterval() int64 {
 	return atomic.LoadInt64(&ssMap.sysVars.refreshInterval)
 }
 
+// SetHistorySize sets the history size for all summaries.
+func (ssMap *stmtSummaryByDigestMap) SetHistorySize(value string, inSession bool) {
+	ssMap.sysVars.Lock()
+	if inSession {
+		ssMap.sysVars.sessionHistorySize = value
+	} else {
+		ssMap.sysVars.globalHistorySize = value
+	}
+	sessionHistorySize := ssMap.sysVars.sessionHistorySize
+	globalHistorySize := ssMap.sysVars.globalHistorySize
+	ssMap.sysVars.Unlock()
+
+	// Calculate the cached `historySize`.
+	size := -1
+	var err error
+	if ssMap.isSet(sessionHistorySize) {
+		size, err = strconv.Atoi(sessionHistorySize)
+		if err != nil {
+			size = -1
+		}
+	}
+	if size < 0 {
+		size, err = strconv.Atoi(globalHistorySize)
+		if err != nil {
+			size = -1
+		}
+	}
+	// Users may set invalid values in configuration file.
+	if size < 0 {
+		size = 0
+	}
+	atomic.StoreInt32(&ssMap.sysVars.historySize, int32(size))
+}
+
+// historySize gets the history size for summaries.
+func (ssMap *stmtSummaryByDigestMap) historySize() int {
+	return int(atomic.LoadInt32(&ssMap.sysVars.historySize))
+}
+
 // newStmtSummaryByDigest creates a stmtSummaryByDigest from StmtExecInfo.
-func newStmtSummaryByDigest(sei *StmtExecInfo, beginTime int64) *stmtSummaryByDigest {
+func newStmtSummaryByDigest(sei *StmtExecInfo, beginTime int64, historySize int) *stmtSummaryByDigest {
 	// Trim SQL to size MaxSQLLength.
 	maxSQLLength := config.GetGlobalConfig().StmtSummary.MaxSQLLength
 	normalizedSQL := sei.NormalizedSQL
 	if len(normalizedSQL) > int(maxSQLLength) {
-		normalizedSQL = normalizedSQL[:maxSQLLength]
+		// Make sure the memory of original `normalizedSQL` will be released.
+		normalizedSQL = string([]byte(normalizedSQL[:maxSQLLength]))
```

Comment:
`		normalizedSQL = normalizedSQL[:maxSQLLength]`

Reply:
No, if the SQL is too long, I want to only save a part of it, not the whole SQL. 
If using slice, the memory of the whole SQL will not freed, leading to memory leak.

Reply:
I got you~